Performing Exploratory Data Analysis (EDA) on a given dataset, you may find some potential issues that need to be addressed before the modeling stage.

here I will introduce you to the issues that you will encounter frequently during your data scientist career (such as duplicated rows, incorrect data types, incorrect values, and missing values) and you will learn about the techniques you can use to easily fix them. But be careful – some issues that you come across don't necessarily need to be fixed. Some of the suspicious or unexpected values you find may be genuine from a business point of view.

It is extremely important to get confirmation either from your stakeholder or the data engineering team before you alter the dataset. It is your responsibility to make sure you are making the right decisions for the business while preparing the dataset.

The Online Retail dataset, which had some negative values in the Quantity column. Here, we expected only positive values. But before fixing this issue straight away (by either dropping the records or transforming them into positive values), it is preferable to get in touch with your stakeholders first and get confirmation that these values are not significant for the business. They may tell you that these values are extremely important as they represent returned items and cost the company a lot of money, so they want to analyze these cases in order to reduce these numbers. If you had moved to the data cleaning stage straight away, you would have missed this critical piece of information and potentially came up with incorrect results.

In [0]:
import pandas as pd

In [0]:
import pandas as pd
file_url = 'https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter10/dataset/Online%20Retail.xlsx?raw=true'
df = pd.read_excel(file_url)

In [3]:
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [4]:
df.tail()

InvoiceNo StockCode  ... CustomerID  Country
541904    581587     22613  ...    12680.0   France
541905    581587     22899  ...    12680.0   France
541906    581587     23254  ...    12680.0   France
541907    581587     23255  ...    12680.0   France
541908    581587     22138  ...    12680.0   France

[5 rows x 8 columns]

In [5]:
df.shape

(541909, 8)

In [6]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [7]:
df.describe().T

count          mean          std  ...       50%       75%      max
Quantity    541909.0      9.552250   218.081158  ...      3.00     10.00  80995.0
UnitPrice   541909.0      4.611114    96.759853  ...      2.08      4.13  38970.0
CustomerID  406829.0  15287.690570  1713.600303  ...  15152.00  16791.00  18287.0

[3 rows x 8 columns]

In [8]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

the True and False binary values correspond to the numerical values 1 and 0, respectively. To find out how many rows have been identified as duplicates, you can use the sum() method on the output of duplicated(). This will add all the 1s (that is, True values) and gives us the count of duplicates

In [9]:
df.duplicated().sum()

5268

 if you want to keep only the variables, that is, InvoiceNo, StockCode, InvoiceDate, and CustomerID, you need to use the following code:

In [10]:
df[['InvoiceNo', 'StockCode', 'InvoiceDate', 'CustomerID']]

InvoiceNo StockCode         InvoiceDate  CustomerID
0         536365    85123A 2010-12-01 08:26:00     17850.0
1         536365     71053 2010-12-01 08:26:00     17850.0
2         536365    84406B 2010-12-01 08:26:00     17850.0
3         536365    84029G 2010-12-01 08:26:00     17850.0
4         536365    84029E 2010-12-01 08:26:00     17850.0
...          ...       ...                 ...         ...
541904    581587     22613 2011-12-09 12:50:00     12680.0
541905    581587     22899 2011-12-09 12:50:00     12680.0
541906    581587     23254 2011-12-09 12:50:00     12680.0
541907    581587     23255 2011-12-09 12:50:00     12680.0
541908    581587     22138 2011-12-09 12:50:00     12680.0

[541909 rows x 4 columns]

If you only want to filter the rows that are considered duplicates, you can use the same API call with the output of the duplicated() method. It will only keep the rows with True as a value:

In [11]:
df[df.duplicated()]

InvoiceNo StockCode  ... CustomerID         Country
517       536409     21866  ...    17908.0  United Kingdom
527       536409     22866  ...    17908.0  United Kingdom
537       536409     22900  ...    17908.0  United Kingdom
539       536409     22111  ...    17908.0  United Kingdom
555       536412     22327  ...    17920.0  United Kingdom
...          ...       ...  ...        ...             ...
541675    581538     22068  ...    14446.0  United Kingdom
541689    581538     23318  ...    14446.0  United Kingdom
541692    581538     22992  ...    14446.0  United Kingdom
541699    581538     22694  ...    14446.0  United Kingdom
541701    581538     23343  ...    14446.0  United Kingdom

[5268 rows x 8 columns]

If you want to subset the rows and columns at the same time, you must use one of the other two available APIs: .loc or .iloc. These APIs do the exact same thing but .loc uses labels or names while .iloc only takes indices as input. You will use the .loc API to subset the duplicated rows and keep only the selected four columns, as shown in the previous example

In [12]:
df.loc[df.duplicated(), ['InvoiceNo', 'StockCode', 'InvoiceDate', 'CustomerID']]

InvoiceNo StockCode         InvoiceDate  CustomerID
517       536409     21866 2010-12-01 11:45:00     17908.0
527       536409     22866 2010-12-01 11:45:00     17908.0
537       536409     22900 2010-12-01 11:45:00     17908.0
539       536409     22111 2010-12-01 11:45:00     17908.0
555       536412     22327 2010-12-01 11:49:00     17920.0
...          ...       ...                 ...         ...
541675    581538     22068 2011-12-09 11:34:00     14446.0
541689    581538     23318 2011-12-09 11:34:00     14446.0
541692    581538     22992 2011-12-09 11:34:00     14446.0
541699    581538     22694 2011-12-09 11:34:00     14446.0
541701    581538     23343 2011-12-09 11:34:00     14446.0

[5268 rows x 4 columns]

This preceding output shows that the first few duplicates are row numbers 517, 527, 537, and so on. By default, pandas doesn't mark the first occurrence of duplicates as duplicates: all the same, duplicates will have a value of True except for the first occurrence. You can change this behavior by specifying the keep parameter. If you want to keep the last duplicate, you need to specify keep='last':

In [13]:
df.loc[df.duplicated(keep = 'last'), ['InvoiceNo', 'StockCode', 'InvoiceDate', 'CustomerID']]

InvoiceNo StockCode         InvoiceDate  CustomerID
485       536409     22111 2010-12-01 11:45:00     17908.0
489       536409     22866 2010-12-01 11:45:00     17908.0
494       536409     21866 2010-12-01 11:45:00     17908.0
521       536409     22900 2010-12-01 11:45:00     17908.0
548       536412     22327 2010-12-01 11:49:00     17920.0
...          ...       ...                 ...         ...
541640    581538     22992 2011-12-09 11:34:00     14446.0
541644    581538     22694 2011-12-09 11:34:00     14446.0
541646    581538     23275 2011-12-09 11:34:00     14446.0
541656    581538     23318 2011-12-09 11:34:00     14446.0
541666    581538     23343 2011-12-09 11:34:00     14446.0

[5268 rows x 4 columns]

As you can see from the previous output, row 485 has the same value as row 539. As expected, row 539 is not marked as a duplicate anymore. If you want to mark all the duplicate records as duplicates, you will have to use keep=False:

In [14]:
df.loc[df.duplicated(keep = False), ['InvoiceNo', 'StockCode', 'InvoiceDate', 'CustomerID']]

InvoiceNo StockCode         InvoiceDate  CustomerID
485       536409     22111 2010-12-01 11:45:00     17908.0
489       536409     22866 2010-12-01 11:45:00     17908.0
494       536409     21866 2010-12-01 11:45:00     17908.0
517       536409     21866 2010-12-01 11:45:00     17908.0
521       536409     22900 2010-12-01 11:45:00     17908.0
...          ...       ...                 ...         ...
541675    581538     22068 2011-12-09 11:34:00     14446.0
541689    581538     23318 2011-12-09 11:34:00     14446.0
541692    581538     22992 2011-12-09 11:34:00     14446.0
541699    581538     22694 2011-12-09 11:34:00     14446.0
541701    581538     23343 2011-12-09 11:34:00     14446.0

[10147 rows x 4 columns]

This time, rows 485 and 539 have been listed as duplicates. Now that you know how to identify duplicate observations, you can decide whether you wish to remove them from the dataset. As we mentioned previously, you must be careful when changing the data.

In the Online Retail dataset, if you take rows 485 and 539 as an example, these two observations are identical. From a business perspective, this means that a specific customer (CustomerID 17908) has bought the same item (StockCode 22111) at the exact same date and time (InvoiceDate 2010-12-01 11:45:00) on the same invoice (InvoiceNo 536409). This is highly suspicious.

In this case, you know that you shouldn't remove these rows. On the other hand, they may tell you that duplication shouldn't happen and that it may be due to human error as the data was entered or during the data extraction step. Let's assume this is the case; now, it is safe for you to remove these rows.

To do so, you can use the drop_duplicates() method from pandas. It has the same keep parameter as duplicated(), which specifies which duplicated record you want to keep or if you want to remove all of them. In this case, we want to keep at least one duplicate row. Here, we want to keep the first occurrence:

In [15]:
df.drop_duplicates(keep='first')

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...    12680.0          France
541905    581587     22899  ...    12680.0          France
541906    581587     23254  ...    12680.0          France
541907    581587     23255  ...    12680.0          France
541908    581587     22138  ...    12680.0          France

[536641 rows x 8 columns]

The output of this method is a new DataFrame that contains unique records where only the first occurrence of duplicates has been kept. If you want to replace the existing DataFrame rather than getting a new DataFrame, you need to use the inplace=True parameter.

The drop_duplicates() and duplicated() methods also have another very useful parameter: subset. This parameter allows you to specify the list of columns to consider while looking for duplicates. By default, all the columns of a DataFrame are used to find duplicate rows. Let's see how many duplicate rows there are while only looking at the InvoiceNo, StockCode, invoiceDate, and CustomerID columns

In [16]:
df.duplicated(subset=['InvoiceNo', 'StockCode', 'InvoiceDate', 'CustomerID'], keep='first').sum()

10677

By looking only at these four columns instead of all of them, we can see that the number of duplicate rows has increased from 5268 to 10677. This means that there are rows that have the exact same values as these four columns but have different values in other columns, which means they may be different records. In this case, it is better to use all the columns to identify duplicate records.